prime-router/docs/authz/authz-design.md

```diff
@@ -0,0 +1,279 @@
+different application frameworks.
+
+### Spring
+
+In Spring we can leverage method security. This allows us to write custom predicates in SpEL (Spring expression language) 
+that will check our access token *before* our application logic runs which will dramatically simplify our code. If this 
+predicate fails then the framework will throw an `AuthorizationDeniedException` which can be caught, logged, and a 
+ReportStream specific 403 response can be returned.
+
+```kotlin
+@PreAuthorize("hasAuthority('SCOPE_org:read') and #oauth2.claim('org').contains(#org)")
+@GetMapping("/api/v1/hello")
+fun hello(org: String): String {
+    return "Hello $org!"
+}
+```
+
+### Azure Functions
+
+Azure functions code will contain a bit more boilerplate code and will have to be done in controller logic. Please note 
+that the below is pseudocode and can be missing important pieces.
+
+```kotlin
+@FunctionName("hello")
+fun hello(
+    @HttpTrigger(
+        name = "hello",
+        methods = [HttpMethod.GET],
+        authLevel = AuthorizationLevel.ANONYMOUS
+    ) request: HttpRequestMessage<String?>,
+    @BindingName("org") org: String,
+): String {
+    if (!authorize(request, "org:read", org)) {
+        return "403"    
+    }
+    
+    return "Hello $org" 
+}
+
+// this would live in some authorization class
+fun authorize(
+    request: HttpRequestMessage,
+    requiredScope: String,
+    maybeRequiredOrg: String?
+): Boolean {
+    val claims = getClaims(request) // pseudocode to grab the claims
+    val containsRequiredScope = claims["scp"].contains(requiredScope)
+    val containsRequiredOrg = maybeRequiredOrg?.let { requiredOrg ->
+        claims["org"].contains(requiredOrg)
+    } ?: true
+    
+    return containsRequiredScope && containsRequiredOrg
+}
+
+/**
+ * Report submission function check
+ * 
+ * Check that a user contains the submit claim
+ * Check that the client header is contained within the appSubmit/userSubmit claim or starts with it (wildcard scenario)
+ *
+ * examples:
+ * submit: [md-phd]
+ * client: md-phd.default
+ * outcome: true
+ * 
+ * submit: [md-phd.default]
+ * client: md-phd.default
+ * outcome: true
+ * 
+ * submit: [md-phd.default]
+ * client: md-phd
+ * outcome: false
+ */
+fun authorizeSubmit(
+    request: HttpRequestMessage,
+): Boolean {
+    val claims = getClaims(request) // pseudocode to grab the claims
+    val containsSubmitScope = claims["scp"].contains("submit")
+    val submitClaim = claims["appSubmit"] ?: claims["userSubmit"] ?: emptyList<String>()
+    val clientHeader = request.headers["client"]
+    
+    val canSubmit = submitClaim.any { sender ->
+        clientHeader.startsWith(sender) || clientHeader == sender
+    }
+    
+    return containsSubmitScope && canSubmit
+}
+
+```
+
+## Deprecated approach to passing groups downstream
+
+At the time of writing this section (2/5/25), code still exists which is doing the following:
```

Comment:
Where is this code? Auth microservice? Also filenames/functions would be helpful.